src/server/transaction.cc

```diff
@@ -798,24 +801,38 @@ void Transaction::ExecuteAsync() {
   auto cb = [seq, this] {
     EngineShard* shard = EngineShard::tlocal();
 
+    // this callback should fail from time to time with non-atomic txs in debug mode because we
+    // clear the shard_data_ array.
+    DCHECK(!this->shard_data_.empty());
+
+    // TODO: to document why we need seqlock_. I suspect I added it to guard shard_data
+    // for multi-command transactions. Specifically, if we resize shard_data_ by coordinator thread
+    // and this callback is executed by engine shard thread, we may access invalid memory.
+    // I think the approach is still broken because we may read seqlock_ and pause, then
+    // the coordinator will progress, exit the WaitForShardCallbacks and will resize shard_data_
+    // and then this callback will call IsArmedInShard and will have invalid memory access.
```

Comment:
True, WaitForShardCallbacks is the following:
```py
await run_count == 0
# <-- gap here
seqlock++
```

so we can increase the seqlock in DecreaseRunCnt instead